core-bundle/src/Resources/contao/library/Contao/Search.php

```diff
@@ -248,26 +248,87 @@ public static function indexPage($arrData)
 			$arrIndex[$strWord] = 1;
 		}
 
+		// Decrement document frequency counts
+		$objDatabase->prepare("
+			UPDATE tl_search_words 
+			INNER JOIN tl_search_index ON tl_search_words.id = tl_search_index.wordId AND tl_search_index.pid = ?
+			SET documentFrequency = GREATEST(0, documentFrequency - 1)
+		")
+			->execute($intInsertId);
+
 		// Remove the existing index
 		$objDatabase->prepare("DELETE FROM tl_search_index WHERE pid=?")
 					->execute($intInsertId);
 
+		// Add new words and increment frequency counts of existing words
+		$objDatabase->prepare("
+			INSERT INTO tl_search_words (word, documentFrequency) 
+			VALUES " . implode(', ', array_fill(0, \count($arrIndex), '(?, 1)')) . "
+			ON DUPLICATE KEY UPDATE documentFrequency = documentFrequency + 1
+		")
+			->execute(array_keys($arrIndex));
+
+		// Remove obsolete words
+		$objDatabase->prepare("DELETE FROM tl_search_words WHERE documentFrequency = 0")
+					->execute();
+
 		$arrQuery = array();
 		$arrValues = array();
 
 		foreach ($arrIndex as $k=>$v)
 		{
-			$arrQuery[] = '(?, ?, ?, ?)';
+			$arrQuery[] = '(?, (SELECT id FROM tl_search_words WHERE word = ?), ?)';
 			$arrValues[] = $intInsertId;
 			$arrValues[] = $k;
 			$arrValues[] = $v;
-			$arrValues[] = $arrData['language'];
 		}
 
 		// Create the new index
-		$objDatabase->prepare("INSERT INTO tl_search_index (pid, word, relevance, language) VALUES " . implode(', ', $arrQuery))
+		$objDatabase->prepare("INSERT INTO tl_search_index (pid, wordId, relevance) VALUES " . implode(', ', $arrQuery))
 					->execute($arrValues);
 
+		list($intMinId, $intMaxId, $intCount) = array_map('intval', $objDatabase->prepare("
+			SELECT IFNULL(MIN(id), 0), IFNULL(MAX(id), 0), COUNT(*) FROM tl_search
+		")->execute()->fetchRow());
```

Comment:
CS looks weird here. 🙈

Reply:
```php
		list($intMinId, $intMaxId, $intCount) = array_map(
			'intval', 
			$objDatabase->prepare("
				SELECT IFNULL(MIN(id), 0), IFNULL(MAX(id), 0), COUNT(*) FROM tl_search
			")->execute()->fetchRow()
		);
```
or
```php
		list($intMinId, $intMaxId, $intCount) = array_map(
			'intval', 
			$objDatabase->prepare("SELECT IFNULL(MIN(id), 0), IFNULL(MAX(id), 0), COUNT(*) FROM tl_search")
				->execute()
				->fetchRow()
		);
```
?

Reply:
Both still look weird. 😄

This is how it would look like with regular indentation:

```php
list($intMinId, $intMaxId, $intCount) = array_map(
	'intval', 
	$objDatabase
		->prepare("SELECT IFNULL(MIN(id), 0), IFNULL(MAX(id), 0), COUNT(*) FROM tl_search")
		->execute()
		->fetchRow()
);
```

But I would probably go for this:

```php
$row = $objDatabase
	->prepare("SELECT IFNULL(MIN(id), 0), IFNULL(MAX(id), 0), COUNT(*) FROM tl_search")
	->execute()
	->fetchRow();

list($intMinId, $intMaxId, $intCount) = array_map('intval', $row);
```

Reply:
Done in 51f0dd08b289d9f5f39b1351976e80b2122e172b